soc/espressif/common/loader.c

```diff
@@ -207,26 +232,402 @@ void map_rom_segments(uint32_t app_drom_start, uint32_t app_drom_vaddr,
+	REG_SET_BIT(SYSTEM_CPU_PERI_CLK_EN_REG, SYSTEM_CLK_EN_ASSIST_DEBUG);
+	REG_CLR_BIT(SYSTEM_CPU_PERI_RST_EN_REG, SYSTEM_RST_EN_ASSIST_DEBUG);
+	REG_WRITE(ASSIST_DEBUG_CORE_0_RCD_EN_REG,
+		  ASSIST_DEBUG_CORE_0_RCD_PDEBUGEN | ASSIST_DEBUG_CORE_0_RCD_RECORDEN);
+#elif CONFIG_SOC_SERIES_ESP32C3
+	REG_SET_BIT(SYSTEM_CPU_PERI_CLK_EN_REG, SYSTEM_CLK_EN_ASSIST_DEBUG);
+	REG_CLR_BIT(SYSTEM_CPU_PERI_RST_EN_REG, SYSTEM_RST_EN_ASSIST_DEBUG);
+	REG_WRITE(ASSIST_DEBUG_CORE_0_RCD_EN_REG,
+		  ASSIST_DEBUG_CORE_0_RCD_PDEBUGEN | ASSIST_DEBUG_CORE_0_RCD_RECORDEN);
+#elif CONFIG_SOC_SERIES_ESP32C6
+	REG_SET_BIT(PCR_ASSIST_CONF_REG, PCR_ASSIST_CLK_EN);
+	REG_CLR_BIT(PCR_ASSIST_CONF_REG, PCR_ASSIST_RST_EN);
+	REG_WRITE(ASSIST_DEBUG_CORE_0_RCD_EN_REG,
+		  ASSIST_DEBUG_CORE_0_RCD_PDEBUGEN | ASSIST_DEBUG_CORE_0_RCD_RECORDEN);
+#endif
+}
+
+static void check_wdt_reset(void)
+{
+	int wdt_rst = 0;
+	soc_reset_reason_t rst_reas;
+
+	rst_reas = esp_rom_get_reset_reason(0);
+	if (rst_reas == RESET_REASON_CORE_RTC_WDT || rst_reas == RESET_REASON_CORE_MWDT0 ||
+#ifndef CONFIG_SOC_SERIES_ESP32C2
+	    rst_reas == RESET_REASON_CORE_MWDT1 ||
+#endif
+	    rst_reas == RESET_REASON_CPU0_MWDT0 ||
+#if defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32C3) ||                    \
+	defined(CONFIG_SOC_SERIES_ESP32C6)
+	    rst_reas == RESET_REASON_CPU0_MWDT1 ||
+#endif
+	    rst_reas == RESET_REASON_CPU0_RTC_WDT) {
+		ESP_EARLY_LOGW(TAG, "PRO CPU has been reset by WDT.");
+		wdt_rst = 1;
+	}
+
+#if defined(CONFIG_SOC_SERIES_ESP32S3) || defined(CONFIG_SOC_SERIES_ESP32)
+	if (wdt_rst) {
+		wdt_reset_info_dump(0);
+	}
+#endif
+	wdt_reset_cpu0_info_enable();
+}
+
+void config_wdt(void)
+{
+	/*
+	 * At this point, the flashboot protection of RWDT and MWDT0 will have been
+	 * automatically enabled. We can disable flashboot protection as it's not
+	 * needed anymore. If configured to do so, we also initialize the RWDT to
+	 * protect the remainder of the bootloader process.
+	 */
+	wdt_hal_context_t rwdt_ctx = RWDT_HAL_CONTEXT_DEFAULT();
+
+	wdt_hal_write_protect_disable(&rwdt_ctx);
+	wdt_hal_set_flashboot_en(&rwdt_ctx, false);
+	wdt_hal_write_protect_enable(&rwdt_ctx);
+
+	/* Disable MWDT0 flashboot protection. But only after we've enabled the RWDT first so that
+	 * there's not gap in WDT protection.
+	 */
+	wdt_hal_context_t mwdt_ctx = {.inst = WDT_MWDT0, .mwdt_dev = &TIMERG0};
+
+	wdt_hal_write_protect_disable(&mwdt_ctx);
+	wdt_hal_set_flashboot_en(&mwdt_ctx, false);
+	wdt_hal_write_protect_enable(&mwdt_ctx);
+}
+
+#ifdef CONFIG_SOC_SERIES_ESP32
+static void reset_mmu(void)
+{
+	/* completely reset MMU in case serial bootloader was running */
+	Cache_Read_Disable(0);
+	Cache_Flush(0);
+	mmu_init(0);
+
+	/* normal ROM boot exits with DROM0 cache unmasked,
+	 * but serial bootloader exits with it masked.
+	 */
+	DPORT_REG_CLR_BIT(DPORT_PRO_CACHE_CTRL1_REG, DPORT_PRO_CACHE_MASK_DROM0);
+}
+#endif /* CONFIG_SOC_SERIES_ESP32 */
+
+static int hardware_init(void)
```

Comment:
Can we provide only `hardware_init` here and leave the rest of the functions in their original locations? I think it is creating unnecessary clutter in the `loader.c`.
Also I'm not sure if having separate `hardware_init` for each SoC is not better (for readability) as the procedures differ between the chips.